Introduce a moose_map_find to get info which object failed the search in a map

GiudGiud opened the pull request:
I would have namespaced it under Moose but right now it conflicts with libmesh's map_find
if we have a better name  I can try again to name space it but map_find is pretty pretty good

GiudGiud committed:
Add a libmesh_map_find-like macro in MOOSE toL
- avoid the libmesh like error message
- add more context about which object returned this error
will help solve  #20485

GiudGiud committed:
Use new map find in framework

GiudGiud committed:
Use moose_map_find in modules

GiudGiud commented on framework/include/utils/MooseUtils.h:
```diff
@@ -1197,6 +1198,129 @@ isDigits(const std::string & str)
   return std::all_of(str.begin(), str.end(), [](unsigned char c) { return std::isdigit(c); });
 }
+
+/**
```
should I credit the copy-pasting from libmesh::Utility?

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1843774/) on d7c1dc9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25858/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25858/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 9cf16544d3d08bb656e169a75cd3034a4b72867c`

dschwen commented on framework/include/base/Moose.h:
```diff
@@ -279,3 +279,12 @@ class ScopedCommSwapper
+// This macro is useful to get around the deficiencies of the C++ map 'at' and 'find' operators
+// at errors with an exception which requires catching to output nicely
+// find is a very verbose syntax
+// This should only be used in a class where name() and type() are defined. If not, rely on
+// libmesh_map_find instead
+#define moose_map_find(map, key)                                                                   \
+  MooseUtils::map_find((map), (key), name(), type(), __FILE__, __LINE__)
```
We can probably do that not as a macro but as a templated function with SFINAE   (passing in `this` - if you want you can wrap that in a macro). Then we can use it anywhere we like and get name/type if it's available, and fall back to vanilla libmesh_find if not.

GiudGiud replied:
Alright gave it a go.

I cant wait to try C++20 for this, SFINAE is heinous to work with

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1844397/) on f1cf4b0 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25858/site)

This comment will be updated on new commits.

GiudGiud committed:
Update description and signature for using SFINAE to recognize a potential class
that defines the type/name operators needed

GiudGiud committed:
Use SFINAE to detect if we are using the map_find macro in a class that defines type/name
If not rely on the libmesh map_find utility
This makes the code more complicated and less maintainable so these macros better be perfect

GiudGiud committed:
Use moose_map_find in more classes

dschwen commented on framework/include/utils/MooseUtils.h:
```diff
@@ -1197,6 +1198,96 @@ isDigits(const std::string & str)
+
+/**
+ * This function should not be called directly (although it can be),
+ * instead see the moose::map_find() macro.
```
```suggestion
 * instead see the moose_map_find() macro.
```

dschwen commented on framework/include/utils/MooseUtils.h:
```diff
@@ -1197,6 +1198,96 @@ isDigits(const std::string & str)
+  return it->second;
+}
+
+/**
+ * Version where this is being used outside of a class so `this` does not exist
+ * First template for non-const returns
+ * Second template for const returns
+ * We enforce that the const-ness of the Map matches the constness of the return argument
+ */
```
I don't get this. Who calls this?

GiudGiud replied:
this is called by the moose_map_find macro if the above substitution failed

dschwen replied:
ok, after your push it makes lightly more sense, but you have the enable ifs backwards I think

dschwen replied:
Nonetheless, if `this` does not exist the macro substitution will give you a non-existing symbol right away.

GiudGiud replied:
yeah. Not much I can do about this?

dschwen commented on framework/include/utils/MooseUtils.h:
```diff
@@ -1197,6 +1198,96 @@ isDigits(const std::string & str)
+    typename T,
+    typename T::type,
+    typename T::name,
+    typename std::enable_if<!libMesh::Utility::is_streamable<Key>::value, Key>::type * = nullptr>
```
So why check if the key is streamable if you never steam it anywhere anyways?

GiudGiud replied:
maybe I forgot to stream it but it should be

dschwen replied:
It looks like this template is only enable if the key is _not_ streamable

GiudGiud committed:
Update framework/include/utils/MooseUtils.h
Add streaming of the key
Flip signs on the streamable

Co-authored-by: Daniel Schwen <daniel@schwen.de>

dschwen commented:
What do you think about making it a member of `MooseObject`? And then have a separate `Moose::findMap(...)` function for use outside of MooseObjects?

GiudGiud commented:
MooseObject and Action ?
That's how I had it initially.

dschwen commented:
> It s going to be a lot of definitions of the same thing
No, you could have all call out to the same helper function in the namespace. Just make the object pointer `nullptr` when you want top avoid calling out to name() / type().

dschwen commented:
You could also introduce a common interface that both MooseObject and Action inherit from and pass a pointer to that.

dschwen commented:
See https://github.com/libMesh/libmesh/pull/3680

GiudGiud commented:
Considering my other PRs trying to make the code base clearer, this is going in the wrong direction.
